sql/core/src/main/scala/org/apache/spark/sql/ArrowConverters.scala

```diff
@@ -0,0 +1,432 @@
+      // case TimestampType => new ArrowType.Timestamp(TimeUnit.MILLISECOND)
+      case _ => throw new UnsupportedOperationException(s"Unsupported data type: $dataType")
+    }
+  }
+
+  /**
+   * Iterate over InternalRows and convert to an ArrowRecordBatch.
+   */
+  private def internalRowIterToArrowBatch(
+      rowIter: Iterator[InternalRow],
+      schema: StructType,
+      allocator: RootAllocator): ArrowRecordBatch = {
+
+    val columnWriters = schema.fields.zipWithIndex.map { case (field, ordinal) =>
+      ColumnWriter(ordinal, allocator, field.dataType)
+        .init()
+    }
+
+    val writerLength = columnWriters.length
+    while (rowIter.hasNext) {
+      val row = rowIter.next()
+      var i = 0
+      while (i < writerLength) {
+        columnWriters(i).write(row)
+        i += 1
+      }
+    }
+
+    val (fieldNodes, bufferArrays) = columnWriters.map(_.finish()).unzip
+    val buffers = bufferArrays.flatten
+
+    val rowLength = if (fieldNodes.nonEmpty) fieldNodes.head.getLength else 0
+    val recordBatch = new ArrowRecordBatch(rowLength,
+      fieldNodes.toList.asJava, buffers.toList.asJava)
+
+    buffers.foreach(_.release())
+    recordBatch
+  }
+
+  /**
+   * Convert a Spark Dataset schema to Arrow schema.
+   */
+  private[sql] def schemaToArrowSchema(schema: StructType): Schema = {
+    val arrowFields = schema.fields.map { f =>
+      new Field(f.name, f.nullable, sparkTypeToArrowType(f.dataType), List.empty[Field].asJava)
+    }
+    new Schema(arrowFields.toList.asJava)
+  }
+
+  /**
+   * Write an ArrowPayload to a byte array
+   */
+  private[sql] def payloadToByteArray(payload: ArrowPayload, schema: StructType): Array[Byte] = {
+    val arrowSchema = ArrowConverters.schemaToArrowSchema(schema)
+    val out = new ByteArrayOutputStream()
+    val writer = new ArrowWriter(Channels.newChannel(out), arrowSchema)
+
+    // Iterate over payload batches to write each one, ensure all batches get closed
+    var batch: ArrowRecordBatch = null
+    Utils.tryWithSafeFinallyAndFailureCallbacks {
+      while (payload.hasNext) {
+        batch = payload.next()
+        writer.writeRecordBatch(batch)
+        batch.close()
+      }
+    }(catchBlock = {
+      Option(batch).foreach(_.close())
+      payload.foreach(_.close())
+    }, finallyBlock = writer.close())
+
+    out.toByteArray
+  }
+}
+
+private[sql] trait ColumnWriter {
+  def init(): this.type
+  def write(row: InternalRow): Unit
+
+  /**
+   * Clear the column writer and return the ArrowFieldNode and ArrowBuf.
+   * This should be called only once after all the data is written.
+   */
+  def finish(): (ArrowFieldNode, Array[ArrowBuf])
+}
+
+/**
+ * Base class for flat arrow column writer, i.e., column without children.
+ */
+private[sql] abstract class PrimitiveColumnWriter(
+  val ordinal: Int,
+  val allocator: BaseAllocator)
+    extends ColumnWriter {
+  def valueVector: BaseDataValueVector
+  def valueMutator: BaseMutator
+
+  def setNull(): Unit
```

Comment:
Yes, but unfortunately it needs to call `valueMutator.setNull()` that is only a member of the concrete Mutator class, so I can't abstract that out. I think on the arrow side it could be defined as an interface and then I could clean it up here.